crypto/engine/eng_cnf.c

```diff
@@ -40,14 +40,38 @@ static int int_engine_init(ENGINE *e)
     return 1;
 }
 
+static ENGINE *int_engine_dynamic_load(const char *engine_id, const char *dynamic_path)
+{
+    ENGINE *e;
+
+    e = ENGINE_by_id("dynamic");
+    if (e == NULL)
+        goto err;
+    if (!ENGINE_ctrl_cmd_string(e, "SO_PATH", dynamic_path, 0))
+        goto err;
+    if (!ENGINE_ctrl_cmd_string(e, "ID", engine_id, 0))
```

Comment:
It is possible that engine_id is NULL here.  It is probably better not to call in this case.

Reply:
I don't think it can happen, see line 78 - `name = skip_dot(name);`, and there is even a debug print of the name in it.

Reply:
oh, but then I dont see why you do this at all.
even this example does not need this complication:
```
[engine_section]
ibmca = ibmca_section

[ibmca_section]
# The openssl engine path for ibmca.so.
# Set the dynamic_path to where the ibmca.so engine
# resides on the system.
dynamic_path = /usr/local/lib/ibmca.so
engine_id = ibmca
init = 1
```
the name is ibmca from the beginning and indeed the
engine_id = ibmca is superfluous but benign.

Reply:
On the other hand, our configure file could do something like that:
```
openssl_conf = ossl_cnf

[ ossl_cnf ]
engines = eng_cnf

[ eng_cnf ]

test = xxx

[ xxx ]
a.dynamic_path = /usr/local/lib/lib1.so
b.dynamic_path = /usr/local/lib/lib2.so
c.dynamic_path = /usr/local/lib/lib3.so
```
works now, but breaks with this patch.

Reply:
Thanks for feedback, I went over your examples, alongside additional configuration files, and now I better understand the configuration types available in the .conf file.

If split to cases, we have the following cases and current (before my fix) behavior:
1. A single engine, using only "engine_id"
2. A single engine, using only "dynamic_path"
3. A single engine, using "engine_id" and then "dynamic_path" - engine_id won't be used at all
4. A single engine, using "dynamic_path" and then "engine_id" - engine_id won't be used at all
5. Multiple engines in same section, using only "dynamic_path"
6. Multiple engines in same section - "engine_id" would only be used if first, and init/ctrl happens **before** the rest of the engines

The init sequence and refcount tracking of the engines makes me believe that either there are bugs when multiple engines (5 and 6) are used in the same section, or this case wasn't meant to be supported at all. The init/ctrl will only work if they occur directly after every engine, and the chances for undefined init behavior are very high. In addition, the reference counting for such engines would be flawed (too high) as only a single engine will get decremented at the end of the function.

All in all, one can load multiple engines, but it is hard to say it will work well in regards to refcounts and init of the engines, and with very high probability "engine_id" (if passed) won't ever be used. I agree that in such cases, "engine_id" isn't well defined anyway, and shouldn't be used.

In the first case, we know the "engine_id", and "ENGINE_by_id()" will use the "ID" command for the dynamic load. Hence, the dynamic load could easily check the engine by name and avoid duplications.

In the second case, we can't really "know" in advance what will be the engine ID, and I agree the dynamic load shouldn't use the "ID" cmd.

In all other cases, "engine_id" won't ever be used, unless the flow is "engine_id", "init", "dynamic_path", and I don't think it was ever meant to be supported. Hence, I argue that in cases 3 and 4, "engine_id" is actually used by users as a "hint" and we could use it when loading the engine by supplying the "ID" command.

If going according to this logic, the only case in which we can insert errors, is in cases 3 and 4: if for some reason there is a mismatch between the "engine_id" and the one of the real loaded engine, and I don't find it very likely. In the multi-load scenario (in a single section), or in case 2, the only thing we lose is the protection against double-load. As the entire fix is a best-effort mechanism, and considering the circumstances, I think it is something we can live with.

Reply:
According to the documentation (https://www.openssl.org/docs/man1.1.1/man5/config.html) each section is an "engine section", and only a single engine is supported per such section. This means the example you used above isn't supported, and thus cases 5 and 6 weren't supported even before my fix.

Again, this makes sense in regards to the init logic, and to the reference count issues I pointed to above.

According to the same documentation, the engine name is indeed set by the configuration section (again, singularity), and can be overriden by "engine_id". Hence, in case 2 we still know what the engine ID should be, and the dynamic engine could use it for the check.

Reply:
I start to agree with @petrovr 
A perfectly valid configuration will just use "dynamic_path", and not necessarily have an "engine_id" at all.
The section name may or may not be identical with the engine id, this name is currently just ignored.
Now this configuration will start to fail, while other really invalid configurations will still fail, but more gracefully.